cohortextractor/query_engines/base_sql.py

```diff
@@ -64,17 +67,16 @@ def get_primary_table(query):
 
 class BaseSQLQueryEngine(BaseQueryEngine):
 
-    sqlalchemy_dialect = NotImplemented
+    sqlalchemy_dialect: Union[type[Dialect], ModuleType]
 
-    custom_types = {}
+    custom_types: dict[str, type[sqlalchemy.types.TypeDecorator]] = {}
     type_map = None
 
     # No limit by default although some DBMSs may impose one
-    max_rows_per_insert = None
+    max_rows_per_insert: Optional[int] = None
 
     def __init_subclass__(cls, **kwargs):
         super().__init_subclass__(**kwargs)
-        assert cls.sqlalchemy_dialect != NotImplemented
```

Comment:
Could this be a truthiness check instead?

Reply:
I thought originally mypy would be checking it for us, but yes a runtime check just to be extra sure can't hurt.

Reply:
No! If mypy _can_ check it then please don't duplicate that at runtime. I'm not sure that it can though, can it?

Reply:
Clearly I made an assumption here! I removed SparkQueryEngine.sqlalchemy_dialect and mypy didn't throw an error, which I really didn't expect…

I'll reinstate this assert until I can work out why mypy isn't catching that.

Reply:
If we make BaseQueryEngine a Protocol we can do this but I'm keen to get these changes in as-is so mypy is configured then tackle the Protocol separately.  I've done a quick spike on switching to a Protocol and it's more work that I expected, so I'll tackle that in a separate PR.

Reply:
:+1: